db/sync.go

```diff
@@ -23,9 +23,22 @@ type SyncInfo struct {
 	ClientViewInfo servetypes.ClientViewInfo `json:"clientViewInfo"`
 }
 
-// BeginSync pushes pending mutations to the data layer and pulls new state via the client view.
-func (db *DB) BeginSync(batchPushURL string, diffServerURL string, diffServerAuth string, dataLayerAuth string, l zl.Logger) (hash.Hash, SyncInfo, error) {
-	syncInfo := SyncInfo{}
+// BeginSync initiates the sync process, temporarily forking the cache
+// onto a new branch called the "sync head", pushing pending mutations
+// to the data layer, and pulling any new state from the data layer via
+// diff-server.
+//
+// If a non-zero syncHead is returned, then caller should call
+// MaybeEndSync (potentially multiple times in a loop) to finalize
+// the sync. See MaybeEndSync for details.
+//
+// Informational details about the push and pull requests are returned
+// via SyncInfo.
+//
+// Returns an error (and zeros for other return values) in the case of
+// invalid argument values, or internal errors.
```

Comment:
if we want this to be more true we should probably adjust it so we return a zero syncinfo if pendingcommits returns an error, or if basesnapshot does, as unlikely as that may be. also push stuffs internal errors into the info rather than returning them. we could make it work more like beginsync now does in order to surface them from sync.